dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,428 @@
+   *
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   * let documentRef = firestore.collection('col').doc();
+   *
+   * bulkWriter
+   *  .set(documentRef, {foo: 'bar'})
+   *  .then(result => {
+   *    console.log('Successfully executed write at: ', result);
+   *  })
+   *  .catch(err => {
+   *    console.log('Write failed with: ', err);
+   *  });
+   * });
+   */
+  set(
+    documentRef: DocumentReference,
+    data: DocumentData,
+    options?: SetOptions
+  ): Promise<WriteResult> {
+    const bulkCommitBatch = this.getEligibleBatch(documentRef);
+    bulkCommitBatch.refsInBatch.add(documentRef);
+    bulkCommitBatch.writeBatch.set(documentRef, data, options);
+
+    return this.getResult(
+      bulkCommitBatch.tag,
+      bulkCommitBatch.writeBatch.opCount
+    );
+  }
+
+  /**
+   * Update fields of the document referred to by the provided
+   * [DocumentReference]{@link DocumentReference}. If the document doesn't yet
+   * exist, the update fails and the entire batch will be rejected.
+   *
+   * The update() method accepts either an object with field paths encoded as
+   * keys and field values encoded as values, or a variable number of arguments
+   * that alternate between field paths and field values. Nested fields can be
+   * updated by providing dot-separated field path strings or by providing
+   * FieldPath objects.
+   *
+   * A Precondition restricting this update can be specified as the last
+   * argument.
+   *
+   * @param {DocumentReference} documentRef A reference to the document to be
+   * updated.
+   * @param {UpdateData|string|FieldPath} dataOrField An object containing the
+   * fields and values with which to update the document or the path of the
+   * first field to update.
+   * @param {...(Precondition|*|string|FieldPath)} preconditionOrValues - An
+   * alternating list of field paths and values to update or a Precondition to
+   * restrict this update
+   * @returns {Promise<WriteResult>} A promise that resolves with the result of
+   * the write. Throws an error if the write fails.
+   *
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   * let documentRef = firestore.collection('col').doc();
+   *
+   * bulkWriter
+   *  .update(documentRef, {foo: 'bar'})
+   *  .then(result => {
+   *    console.log('Successfully executed write at: ', result);
+   *  })
+   *  .catch(err => {
+   *    console.log('Write failed with: ', err);
+   *  });
+   * });
+   */
+  update(
+    documentRef: DocumentReference,
+    dataOrField: UpdateData | string | FieldPath,
+    ...preconditionOrValues: Array<
+      {lastUpdateTime?: Timestamp} | unknown | string | FieldPath
+    >
+  ): Promise<WriteResult> {
+    const bulkCommitBatch = this.getEligibleBatch(documentRef);
+    bulkCommitBatch.refsInBatch.add(documentRef);
+    bulkCommitBatch.writeBatch.update(
+      documentRef,
+      dataOrField,
+      preconditionOrValues
+    );
+
+    return this.getResult(
```

Comment:
What do you think of making a WriteBatch that returns a Promise for each operation? This should simplify your tracking here and remove some of the state from here.

You can take a look at what Java does: https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/firestore/UpdateBuilder.html

This is a super class that Transaction and WriteBatch use. You could use an UpdateBuilder that returns Promises, and then ignore this in the WriteBatch class.